src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+        """
+
+        # verify block_size is even
+        block_size = block_size - (block_size % 2)
+
+        size_x = micrograph.shape[1]
+        size_y = micrograph.shape[0]
+
+        step_size = block_size // 2
+        range_y = size_y // step_size - 1
+        range_x = size_x // step_size - 1
+
+        block_list = [
+            micrograph[
+                i * step_size : (i + 2) * step_size, j * step_size : (j + 2) * step_size
+            ]
+            for j in range(range_y)
+            for i in range(range_x)
+        ]
+        block = np.asarray(block_list, dtype=micrograph.dtype)
+
+        block_sum = np.sum(np.sum(block, axis=-1), axis=-1)
+        block_sum = np.reshape(block_sum, (block.shape[0], 1, 1))
+        block_sum = np.tile(block_sum, (1, block_size, block_size))
+        block = block - (
+            block_sum / (block_size ** 2)
+        )  # equals to the matlab version (11-7-19)
+
+        return block
+
+    def ctf_tapers(self, N, R, L):
+        """
+        Compute data tapers (which are discrete prolate spheroidal sequences (dpss))
+
+        :param N: Size of each taper
+        :param R: Spectral resolution
+        :param L: Number of tapers
+        :return: numpy array of data tapers
+        """
+
+        k, el = np.meshgrid(
+            np.arange(N, dtype=self.dtype), np.arange(N, dtype=self.dtype)
+        )
+
+        denom = np.pi * (k - el)
+        denom = denom + np.eye(N, dtype=self.dtype)
+        phi_R = np.divide(np.sin(np.pi * R * (k - el)), denom)
+        np.fill_diagonal(phi_R, 1)  # absolute difference from Matlab 10^-18
+
+        data_tapers, _ = eigs(phi_R, L)
+
+        return data_tapers
+
+    def ctf_estimate_psd(self, blocks, tapers_1d, num_1d_tapers):
+        """
+        Estimate the power spectrum of the micrograph using the multi-taper method
+
+        :param blocks: 3-D numpy array containing windows extracted from the micrograph in the ctf_preprocess function.
+        :param tapers_1d: numpy array of data tapers.
+        :param num_1d_tapers: number of 1-D data tapers
+        :return: Numpy array of estimated power spectrum.
+        """
+
+        tapers_1d = tapers_1d.astype(complex_type(self.dtype), copy=False)
+
+        blocks_mt_pre_fft = np.zeros(
+            blocks[0, :, :].shape, dtype=complex_type(self.dtype)
+        )
+
+        blocks_mt_post_fft = np.zeros(
+            blocks_mt_pre_fft.shape, dtype=complex_type(self.dtype)
+        )
+
+        blocks_mt = np.zeros(blocks_mt_pre_fft.shape, dtype=self.dtype)
+
+        blocks_tapered = np.zeros(blocks[0, :, :].shape, dtype=complex_type(self.dtype))
+
+        taper_2d = np.zeros(
+            (blocks.shape[1], blocks.shape[2]), dtype=complex_type(self.dtype)
+        )
+
+        for x in range(num_1d_tapers ** 2):
+            np.matmul(
+                np.reshape(tapers_1d[:, x // num_1d_tapers], (tapers_1d.shape[0], 1)),
```

Comment:
Replace with grid and was still ugly. i ended up with nothing fancy:

```
+        for ax1 in range(num_1d_tapers):
+            for ax2 in range(num_1d_tapers):
+                np.matmul(
+                    tapers_1d[:, ax1, np.newaxis],
+                    tapers_1d[:, ax2, np.newaxis].T,
+                    out=taper_2d,
+                )
```

Probably could turn that into a bigger matmul, but not sure it would matter (looking at next block).

Reply:
Eh but it's easier to read, at least to my eyes.